src/vm/types/serialization.rs

```diff
@@ -0,0 +1,884 @@
+                } else {
+                    Err(InterpreterError::DeserializeUnexpectedTypeField(type_n).into())
+                }
+            },
+            JSONParser::None { type_n } => {
+                check_match!(expected_type, AtomTypeIdentifier::OptionalType(_))?;
+                if type_n == TYPE_OPT_NONE {
+                    Ok(Value::none())
+                } else {
+                    Err(InterpreterError::DeserializeUnexpectedTypeField(type_n).into())
+                }
+            },
+            JSONParser::Container { type_n, value } => {
+                let outer_type = match type_n.as_str() {
+                    TYPE_RESP_OK => Ok(ContainerTypes::RESP_OK),
+                    TYPE_RESP_ERR => Ok(ContainerTypes::RESP_ERR),
+                    TYPE_OPT_SOME => Ok(ContainerTypes::OPT_SOME),
+                    _ => Err(InterpreterError::DeserializeUnexpectedTypeField(type_n))
+                }?;
+
+                let expect_contained_type = match expected_type {
+                    None => None,
+                    Some(x) => {
+                        let contained_type = match (&outer_type, x) {
+                            (ContainerTypes::RESP_OK,  TypeSignature::Atom(AtomTypeIdentifier::ResponseType(types))) => Ok(&types.0),
+                            (ContainerTypes::RESP_ERR, TypeSignature::Atom(AtomTypeIdentifier::ResponseType(types))) => Ok(&types.1),
+                            (ContainerTypes::OPT_SOME, TypeSignature::Atom(AtomTypeIdentifier::OptionalType(some_type))) => Ok(some_type.as_ref()),
+                            _ => Err(InterpreterError::DeserializeExpected(x.clone()))
+                        }?;
+                        Some(contained_type)
+                    }
+                };
+
+                let deserialized_value = Value::try_deserialize_parsed(*value, expect_contained_type)?;
+                match outer_type {
+                    ContainerTypes::RESP_OK => Ok(Value::okay(deserialized_value)),
+                    ContainerTypes::RESP_ERR => Ok(Value::error(deserialized_value)),
+                    ContainerTypes::OPT_SOME => Ok(Value::some(deserialized_value))
+                }
+            },
+            JSONParser::List { type_n, mut entries } => {
+                if type_n != TYPE_LIST {
+                    return Err(InterpreterError::DeserializeUnexpectedTypeField(type_n).into())
+                }
+
+                let (list_type, entry_type) = match expected_type {
+                    None => (None, None),
+                    Some(TypeSignature::List(list_type)) => (Some(list_type), Some(list_type.get_list_item_type())),
+                    Some(x) => return Err(InterpreterError::DeserializeExpected(x.clone()).into())
+                };
+
+                let items: InterpreterResult<_> = entries
+                    .drain(..)
+                    .map(|value| Value::try_deserialize_parsed(value, entry_type.as_ref()))
```

Comment:
Should this method bound its recursion, i.e. as a safety/sanity check?

Reply:
I don't think so -- any recursion happening here has already needed to perform the same recursion during the construction of the value.